File: C#/Stream/StreamBurst/StreamBurst.cs
﻿//-----------------------------------------------------------------------------
// StreamBurst.cs
//
// Demonstrates how to use the StreamBurst function for streaming.
//
// support@labjack.com
//-----------------------------------------------------------------------------
using System;
using System.Diagnostics;
using LabJack;

namespace StreamBurst
{
    class StreamBurst
    {
        static void Main(string[] args)
        {
            StreamBurst sb = new StreamBurst();
            sb.performActions();
        }

        public void showErrorMessage(LJM.LJMException e)
        {
            Console.Out.WriteLine("LJMException: " + e.ToString());
            Console.Out.WriteLine(e.StackTrace);
        }

        public void performActions()
        {
            int handle = 0;
            int devType = 0;
            int conType = 0;
            int serNum = 0;
            int ipAddr = 0;
            int port = 0;
            int maxBytesPerMB = 0;
            string ipAddrStr = "";

            try
            {
                //Open first found LabJack
                LJM.OpenS("ANY", "ANY", "ANY", ref handle);
                //devType = LJM.CONSTANTS.dtANY; //Any device type
                //conType = LJM.CONSTANTS.ctANY; //Any connection type
                //LJM.Open(devType, conType, "ANY", ref handle);

                LJM.GetHandleInfo(handle, ref devType, ref conType, ref serNum, ref ipAddr, ref port, ref maxBytesPerMB);
                LJM.NumberToIP(ipAddr, ref ipAddrStr);
                Console.WriteLine("Opened a LabJack with Device type: " + devType + ", Connection type: " + conType + ",");
                Console.WriteLine("Serial number: " + serNum + ", IP address: " + ipAddrStr + ", Port: " + port + ",");
                Console.WriteLine("Max bytes per MB: " + maxBytesPerMB);

                //Stream Configuration
                uint numScans = 10000; //Number of scans to perform
                string[] aScanListNames = new String[] { "AIN0", "AIN1" }; //Scan list names to stream.
                int numAddresses = aScanListNames.Length;
                int[] aTypes = new int[numAddresses]; //Dummy
                int[] aScanList = new int[numAddresses]; //Scan list addresses to stream. StreamBurst uses Modbus addresses.
                LJM.NamesToAddresses(numAddresses, aScanListNames, aScanList, aTypes);
                double scanRate = 5000; //Scans per second
                double[] aData = new double[numScans * numAddresses];

                try
                {
                    //Configure the analog inputs' negative channel, range, settling time and
                    //resolution.
                    //Note when streaming, negative channels and ranges can be configured for
                    //individual analog inputs, but the stream has only one settling time and
                    //resolution.
                    string[] aNames = new string[] { "AIN_ALL_NEGATIVE_CH", "AIN_ALL_RANGE", "STREAM_SETTLING_US", "STREAM_RESOLUTION_INDEX" };
                    double[] aValues = new double[] { LJM.CONSTANTS.GND, 10.0, 0, 0 };  //single-ended, +/-10V, 0 (default), 0 (default)
                    int errorAddress = 0;
                    LJM.eWriteNames(handle, aNames.Length, aNames, aValues, ref errorAddress);

                    Console.WriteLine("\nScan list:");
                    for(int i = 0; i < numAddresses; i++)
                        Console.WriteLine("  " + aScanListNames[i]);
                    Console.WriteLine("Scan rate = " + scanRate + " Hz");
                    Console.WriteLine("Sample rate = " + (scanRate * numAddresses) + " Hz");
                    Console.WriteLine("Number of scans = " + numScans);
                    Console.WriteLine("Seconds of samples = " + (numScans / scanRate) + " seconds");

                    Console.Write("\nStreaming with StreamBurst...");
                    
                    Stopwatch sw = new Stopwatch();
                    sw.Start();
                    
                    //Stream data using StreamBurst
                    LJM.StreamBurst(handle, numAddresses, aScanList, ref scanRate, numScans, aData);
                    
                    sw.Stop();
                    
                    Console.WriteLine(" Done");

                    //Count the skipped samples which are indicated by -9999 values. Missed
                    //samples occur after a device's stream buffer overflows and are reported
                    //after auto-recover mode ends.
                    ulong skippedTotal = 0;
                    foreach(double d in aData)
                    {
                        if (d == -9999.00)
                            skippedTotal++;
                    }

                    Console.WriteLine("\nSkipped scans = " + (skippedTotal / (ulong)numAddresses));
                    double time = sw.ElapsedMilliseconds / 1000.0;
                    Console.WriteLine("Time taken = " + time + " seconds");

                    Console.WriteLine("\nLast scan:");
                    for (int i = 0; i < numAddresses; i++)
                        Console.WriteLine("  " + aScanListNames[i] + " = " + aData[(numScans-1)*numAddresses + i]);
                }
                catch (LJM.LJMException e)
                {
                    showErrorMessage(e);
                }
            }
            catch (LJM.LJMException e)
            {
                showErrorMessage(e);
            }

            LJM.CloseAll(); //Close all handles

            Console.WriteLine("\nDone.\nPress the enter key to exit.");
            Console.ReadLine(); // Pause for user
        }
    }
}

File: VB/Stream/StreamBurst/StreamBurst.vb
'------------------------------------------------------------------------------
' StreamBurst.vb
'
' Demonstrates how to use the StreamBurst function for streaming.
'
' support@labjack.com
'------------------------------------------------------------------------------
Option Explicit On

Imports LabJack
Imports System.Diagnostics


Module StreamBurst

    Sub showErrorMessage(ByVal e As LJM.LJMException)
        Console.WriteLine("LJMException: " & e.ToString)
        Console.WriteLine(e.StackTrace)
    End Sub

    Sub displayHandleInfo(ByVal handle As Integer)
        Dim devType As Integer
        Dim conType As Integer
        Dim serNum As Integer
        Dim ipAddr As Integer
        Dim port As Integer
        Dim maxBytesPerMB As Integer
        Dim ipAddrStr As String = ""

        LJM.GetHandleInfo(handle, devType, conType, serNum, ipAddr, port, _
                          maxBytesPerMB)
        LJM.NumberToIP(ipAddr, ipAddrStr)
        Console.WriteLine("Opened a LabJack with Device type: " & devType & _
                          ", Connection type: " & conType & ",")
        Console.WriteLine("Serial number: " & serNum & ", IP address: " & _
                          ipAddrStr & ", Port: " & port & ",")
        Console.WriteLine("Max bytes per MB: " & maxBytesPerMB)
    End Sub

    Sub Main()
        Dim handle As Integer
        Dim aNames() As String
        Dim aValues() As Double
        Dim errAddr As Integer = -1

        Dim numScans As Integer
        Dim numAddresses As Integer
        Dim aScanListNames() As String
        Dim aTypes() As Integer
        Dim aScanList() As Integer
        Dim scanRate As Double

        Dim aData() As Double
        Dim skippedTotal As UInt32
        Dim sw As Stopwatch
        Dim time As Double

        Try
            ' Open first found LabJack
            LJM.OpenS("ANY", "ANY", "ANY", handle)  ' Any device, Any connection, Any identifier
            'LJM.OpenS("T7", "ANY", "ANY", handle)  ' T7 device, Any connection, Any identifier
            'LJM.OpenS("T4", "ANY", "ANY", handle)  ' T4 device, Any connection, Any identifier
            'LJM.Open(LJM.CONSTANTS.dtANY, LJM.CONSTANTS.ctANY, "ANY", handle)  ' Any device, Any connection, Any identifier

            displayHandleInfo(handle)

            ' Stream Configuration
            numScans = 20000 ' Number of scans to perform
            numAddresses = 2
            ReDim aScanListNames(numAddresses - 1)  ' Scan list names to stream.
            aScanListNames(0) = "AIN0"
            aScanListNames(1) = "AIN1"
            ReDim aTypes(numAddresses - 1)  ' Dummy
            ReDim aScanList(numAddresses - 1)  ' Scan list addresses to stream. StreamBurst uses Modbus addresses.
            LJM.NamesToAddresses(numAddresses, aScanListNames, aScanList, aTypes)
            scanRate = 10000  ' Scans per second
            ReDim aData(numScans * numAddresses - 1)

            Try
                ' Configure the analog inputs' negative channel, range, settling time and
                ' resolution.
                ' Note when streaming, negative channels and ranges can be configured for
                ' individual analog inputs, but the stream has only one settling time and
                ' resolution.
                ReDim aNames(4)
                aNames(0) = "AIN_ALL_NEGATIVE_CH"
                aNames(1) = "AIN0_RANGE"
                aNames(2) = "AIN1_RANGE"
                aNames(3) = "STREAM_SETTLING_US"
                aNames(4) = "STREAM_RESOLUTION_INDEX"
                ReDim aValues(4)
                aValues(0) = LJM.CONSTANTS.GND  ' single-ended
                aValues(1) = 10.0  ' +/-10V
                aValues(2) = 10.0  ' +/-10V
                aValues(3) = 0  ' 0 = default
                aValues(4) = 0  ' 0 = default
                LJM.eWriteNames(handle, 4, aNames, aValues, errAddr)

                Console.WriteLine("")
                Console.WriteLine("Scan list:")
                For i = 0 To numAddresses - 1
                    Console.WriteLine("  " & aScanListNames(i))
                Next
                Console.WriteLine("Scan rate = " & scanRate & " Hz")
                Console.WriteLine("Sample rate = " & (scanRate * numAddresses) & " Hz")
                Console.WriteLine("Total number of scans = " & numScans)
                Console.WriteLine("Total number of samples = " & (numScans * numAddresses))
                Console.WriteLine("Seconds of samples = " & (numScans / scanRate) & " seconds")

                Console.WriteLine("")
                Console.Write("Streaming with StreamBurst...")

                sw = New Stopwatch()
                sw.Start()

                ' Stream data using StreamBurst
                LJM.StreamBurst(handle, numAddresses, aScanList, scanRate, numScans, aData)

                sw.Stop()

                Console.WriteLine(" Done")

                Console.WriteLine("")
                Console.WriteLine("Skipped scans = " & _
                                  (skippedTotal / numAddresses))
                time = sw.ElapsedMilliseconds / 1000.0
                Console.WriteLine("Time taken = " & time & " seconds")

                Console.WriteLine("")
                Console.WriteLine("Last scan:")
                For i = 0 To numAddresses - 1
                    Console.WriteLine("  " & aScanListNames(i) & " = " & _
                                      aData((numScans - 1) * numAddresses + i))
                Next

            Catch ljme As LJM.LJMException
                showErrorMessage(ljme)
            End Try

        Catch ljme As LJM.LJMException
            showErrorMessage(ljme)
        End Try

        LJM.CloseAll()  ' Close all handles

        Console.WriteLine("")
        Console.WriteLine("Done.")
        Console.WriteLine("Press the enter key to exit.")
        Console.ReadLine()  ' Pause for user
    End Sub

End Module
